File: src/0402-Remove-K-Digits/remove_k_digits.js
/**
 * @param {string} num
 * @param {number} k
 * @return {string}
 */
var removeKdigits = function(num, k) {
  let stack = []
        
  for (let i = 0; i < num.length; i++) {
    while (k > 0 && stack.length > 0 && stack[stack.length - 1] > num[i]) {
      stack.pop()
      k -= 1
    }
    stack.push(num[i])
  }
  
  while (k > 0) {
    stack.pop()
    k -= 1
  }
  
  let result = ''
  let leadingZero = true
  for (let digit of stack) {
    if (leadingZero && digit === '0') {
      continue
    }
    leadingZero = false
    result += digit
  }
  
  return result === '' ? '0' : result
};

const num = "1432219"
const k = 3

console.log(removeKdigits(num, k))

File: src/0402-Remove-K-Digits/remove_k_digits.cpp
#include <iostream>
#include <string>
#include <vector>

class Solution {
public:
    std::string removeKdigits(std::string num, int k) {
        std::vector<char> stack;
        
        for (char n : num) {
            while (k > 0 && !stack.empty() && stack.back() > n) {
                stack.pop_back();
                k -= 1;
            }
            stack.push_back(n);
        }
        
        while (k > 0) {
            stack.pop_back();
            k -= 1;
        }

        std::string ans = "";
        bool leadingZero = true;

        for (char c : stack) {
            if (leadingZero && c == '0') continue;
            leadingZero = false;
            ans += c;
        }
        
        return ans.empty() ? "0" : ans;
    }
};

int main() {
    std::string num = "1432219";
    int k = 3;
    Solution solution;

    std::string result = solution.removeKdigits(num, k);
    
    std::cout << "Result: " << result << std::endl;
    return 0;
}